runtime/src/gpu/rocm/gpu-rocm.c

```diff
@@ -27,67 +27,497 @@
+  bool src_on_host = chpl_gpu_impl_is_host_ptr(src);
+
+  if (!dst_on_host && !src_on_host) {
+    chpl_gpu_impl_copy_device_to_device(dst, src, n);
+    return dst;
+  }
+  else if (!dst_on_host) {
+    chpl_gpu_impl_copy_host_to_device(dst, src, n);
+    return dst;
+  }
+  else if (!src_on_host) {
+    chpl_gpu_impl_copy_device_to_host(dst, src, n);
+    return dst;
+  }
+  else {
+    assert(dst_on_host && src_on_host);
+    return memmove(dst, src, n);
+  }
+  #else
+
+  // for unified memory strategy we don't want to generate calls to copy
+  // data from the device to host (since it can just be accessed directly)
   return memmove(dst, src, n);
+  #endif
 }
 
-void chpl_gpu_impl_hostmem_register(void *memAlloc, size_t size) { }
+void chpl_gpu_impl_copy_device_to_host(void* dst, const void* src, size_t n) {
+  assert(chpl_gpu_is_device_ptr(src));
 
-void chpl_gpu_impl_copy_device_to_host(void* dst, const void* src, size_t n) {}
-void chpl_gpu_impl_copy_host_to_device(void* dst, const void* src, size_t n) {}
+  CUDA_CALL(hipMemcpyDtoH(dst, (hipDeviceptr_t)src, n));
+}
 
-void chpl_gpu_impl_copy_device_to_device(void* dst, const void* src, size_t n) {}
+void chpl_gpu_impl_copy_host_to_device(void* dst, const void* src, size_t n) {
+  assert(chpl_gpu_is_device_ptr(dst));
 
-bool chpl_gpu_impl_is_device_ptr(const void* ptr) {
-  return false;
+  CUDA_CALL(hipMemcpyHtoD((hipDeviceptr_t)dst, src, n));
 }
 
-bool chpl_gpu_impl_is_host_ptr(const void* ptr) {
-  return false;
+void chpl_gpu_impl_copy_device_to_device(void* dst, const void* src, size_t n) {
+  assert(chpl_gpu_is_device_ptr(dst) && chpl_gpu_is_device_ptr(src));
+
+  CUDA_CALL(hipMemcpyDtoD((hipDeviceptr_t)dst, (hipDeviceptr_t)src, n));
+}
+
+
+void* chpl_gpu_impl_comm_async(void *dst, void *src, size_t n) {
+/*  hipStream_t stream;
+  hipStreamCreateWithFlags(&stream, hipStreamNonBlocking);
+  cuMemcpyAsync((hipDeviceptr_t)dst, (hipDeviceptr_t)src, n, stream);
+  return stream;*/
+  assert(false);
+  return NULL;
+}
+
+void chpl_gpu_impl_comm_wait(void *stream) {
+  hipStreamSynchronize((hipStream_t)stream);
+  hipStreamDestroy((hipStream_t)stream);
+}
+
+void* chpl_gpu_mem_array_alloc(size_t size, chpl_mem_descInt_t description,
+                               int32_t lineno, int32_t filename) {
+
+  chpl_gpu_ensure_context();
+
+  CHPL_GPU_DEBUG("chpl_gpu_mem_array_alloc called. Size:%d file:%s line:%d\n", size,
+               chpl_lookupFilename(filename), lineno);
+
+  hipDeviceptr_t ptr = 0;
+  if (size > 0) {
+    chpl_memhook_malloc_pre(1, size, description, lineno, filename);
+#ifdef CHPL_GPU_MEM_STRATEGY_ARRAY_ON_DEVICE
+    if (chpl_gpu_device_alloc) {
+      CUDA_CALL(hipMalloc(&ptr, size));
+    }
+    else {
+//      void* mem = chpl_mem_alloc(size, description, lineno, filename);
+//      CHPL_GPU_DEBUG("\tregistering %p\n", mem);
+//      CUDA_CALL(hipHostRegister(mem, size, hipHostRegisterPortable));
+//      CUDA_CALL(hipHostGetDevicePointer(&ptr, mem, 0));
```

Comment:
another limitation?

Reply:
I'm not totally sure. I recall at some point of time I was running into issues with this but trying it again with `array_on_device` it seems to be working. Anyway I've changed it to this; we can revisit if issues surface:

```cpp
void* mem = chpl_mem_alloc(size, description, lineno, filename);
CHPL_GPU_DEBUG("\tregistering %p\n", mem);
ROCM_CALL(hipHostRegister(mem, size, hipHostRegisterPortable));
ROCM_CALL(hipHostGetDevicePointer(&ptr, mem, 0));
return mem;
```